vaporetto/src/type_scorer.rs

```diff
@@ -78,172 +65,366 @@ impl TypeScorerPma {
-            // Therefore, the following code is safe.
-            let weights = unsafe { self.weights.get_unchecked(m.value()) };
-            weights.add_weight(ys, offset);
-        }
+        self.map
+            .into_iter()
+            .map(|(ngram, weight)| (ngram, weight.into_inner().0))
+            .collect()
     }
 }
 
-impl<'de> BorrowDecode<'de> for TypeScorerPma {
-    /// WARNING: This function is inherently unsafe. Do not publish this function outside this
-    /// crate.
-    fn borrow_decode<D: BorrowDecoder<'de>>(decoder: &mut D) -> Result<Self, DecodeError> {
-        let pma_data: &[u8] = BorrowDecode::borrow_decode(decoder)?;
-        let (pma, _) =
-            unsafe { DoubleArrayAhoCorasick::deserialize_from_slice_unchecked(pma_data) };
-        Ok(Self {
-            pma,
-            weights: Decode::decode(decoder)?,
-            window_size: Decode::decode(decoder)?,
-        })
-    }
-}
+/// WARNING: Decoding is inherently unsafe. Do not publish this struct outside this
+/// crate.
+#[derive(BorrowDecode, Encode)]
+pub enum TypeScorer {
+    Boundary(TypeScorerBoundary),
 
-impl Encode for TypeScorerPma {
-    fn encode<E: bincode::enc::Encoder>(
-        &self,
-        encoder: &mut E,
-    ) -> Result<(), bincode::error::EncodeError> {
-        let pma_data = self.pma.serialize_to_vec();
-        Encode::encode(&pma_data, encoder)?;
-        Encode::encode(&self.weights, encoder)?;
-        Encode::encode(&self.window_size, encoder)?;
-        Ok(())
-    }
-}
+    #[cfg(feature = "cache-type-score")]
+    BoundaryCache(TypeScorerBoundaryCache),
 
-#[cfg(feature = "cache-type-score")]
-#[derive(Decode, Encode)]
-pub struct TypeScorerCache {
-    scores: Vec<i32>,
-    window_size: u8,
-    sequence_mask: usize,
+    #[cfg(feature = "tag-prediction")]
+    BoundaryTag(TypeScorerBoundaryTag),
 }
 
-#[cfg(feature = "cache-type-score")]
-impl TypeScorerCache {
-    pub fn new(model: NgramModel<Vec<u8>>, window_size: u8) -> Result<Self> {
-        let pma = DoubleArrayAhoCorasick::new(model.data.iter().map(|d| &d.ngram))
-            .map_err(|_| VaporettoError::invalid_model("invalid character type n-grams"))?;
-        let mut weights = vec![];
-        for d in model.data {
-            if d.weights.len() <= 2 * usize::from(window_size) - d.ngram.len() {
-                return Err(VaporettoError::invalid_model(
-                    "invalid size of weight vector",
-                ));
+impl TypeScorer {
+    pub fn new(
+        ngram_model: NgramModel<Vec<u8>>,
+        window_size: u8,
+        #[cfg(feature = "tag-prediction")] tag_ngram_model: Vec<TagNgramModel<Vec<u8>>>,
+    ) -> Result<Self> {
+        #[cfg(feature = "tag-prediction")]
+        if tag_ngram_model.is_empty() {
+            match window_size {
+                #[cfg(feature = "cache-type-score")]
+                0..=3 => Ok(Self::BoundaryCache(TypeScorerBoundaryCache::new(
```

Comment:
I guess that the maximum window size `3` should be defined with a constant variable because the value is used multiple times.